pkg/metadata/store/metadata_store.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package store
+
+import (
+	"fmt"
+	"sync"
+)
+
+// All byte arrays are expected to be read-only. User MUST NOT modify byte
+// array element directly!!
+
+// UpdateFunc is function used to update a specific metadata. The value
+// passed in is the old value. The updated value should be returned.
+// If there is an error, the update will be rolled back.
+type UpdateFunc func([]byte) ([]byte, error)
+
+// MetadataStore is the interface for storing metadata.
+// TODO(random-liu): Initialize the metadata store with a type, and replace
+// []byte with interface{}, so as to avoid extra marshal/unmarshal on the
+// user side.
+type MetadataStore interface {
+	// Create the metadata containing the passed in data with the
+	// specified id.
+	// Note:
+	// * Create MUST return error if the id already exists.
+	// * The id and data MUST be added in one transaction to the store.
+	Create(string, []byte) error
+	// Get the data by id.
+	// Note that Get MUST return nil without error if the id
+	// doesn't exist.
+	Get(string) ([]byte, error)
+	// Update the data by id. Note that the update MUST be applied in
+	// one transaction.
+	Update(string, UpdateFunc) error
+	// List returns entire array of data from the store.
+	List() ([][]byte, error)
+	// Delete the data by id.
+	// Note:
+	// * Delete should be idempotent, it MUST not return error if the id
+	// doesn't exist or has been removed.
+	// * The id and data MUST be deleted in one transaction.
+	Delete(string) error
+}
+
+// TODO(random-liu) Add checkpoint. When checkpoint is enabled, it should cache data
+// in memory and checkpoint metadata into files during update. Metadata should serve
+// from memory, but any modification should be checkpointed, so that memory could be
+// recovered after restart. It should be possible to disable the checkpoint for testing.
+// Note that checkpoint update may fail, so the recovery logic should tolerant that.
+
+// An atomic file operator is needed.
```

Comment:
This is a part of TODO. We need atomic file writer to do checkpoint. Could be removed now.